crates/evm-nonces-recovery/src/lib.rs

```diff
@@ -0,0 +1,216 @@
+            "There should be no accounts left for recovery",
+        );
+        Ok(())
+    }
+}
+
+impl<R, Precompiles> MigrationBrokenNoncesRecovery<R, Precompiles>
+where
+    R: pallet_evm::Config,
+    R: pallet_evm_system::Config<AccountId = H160>,
+    Precompiles: TypedGet,
+    Precompiles::Type: IntoIterator<Item = H160>,
+{
+    /// Checks the contract account state and recovers it if necessary.
+    /// - If the state is missing, recreates it.
+    /// - If the state has nonce = 0, writes the smallest possible valid nonce.
+    ///
+    /// Precompiled contracts in Ethereum usually have nonce = 0 as well. However, since precompiled
+    /// contracts are typically implemented by hooking calls to specific addresses and adding dummy
+    /// state (to ensure they are callable like regular contracts), there's no need for a non-zero
+    /// nonce unless they explicitly perform state-changing operations like `CREATE`. Therefore,
+    /// precompiled contracts MUST NOT be passed here.
+    fn recover(id: &H160, account: &mut Option<AccountInfoOf<R>>) -> Weight {
+        let Some(account) = account.as_mut() else {
+            info!("Account {id} lacks its state");
+            let (nonce, weight) = Self::min_nonce(id);
+            *account = Some(AccountInfo {
+                nonce,
+                data: Default::default(),
+            });
+            return weight;
+        };
+        if !account.nonce.is_zero() {
+            return Default::default();
+        }
+        info!("Account {id} has zero nonce");
+        let (nonce, weight) = Self::min_nonce(id);
+        account.nonce = nonce;
+        weight
+    }
+
+    /// Computes the minimum possible nonce for a given account.
+    fn min_nonce(id: &H160) -> (<R as pallet_evm_system::Config>::Index, Weight) {
+        let mut weight = Weight::default();
+        let nonce = (1..)
+            .find(|&nonce| {
+                let contract_id = contract_address(id, nonce);
+                let (is_known_to_evm, w) = Self::is_known_to_evm(&contract_id);
+                weight.saturating_accrue(w);
+                !is_known_to_evm
+                // EVM system may also have a state (pallet_evm_system::Account) behind this
+                // contract_id that MUST NOT cause us to skip the nonce. This state may keep
+                // a deposit made to the address before the contract was created ("pre-funded
+                // contract"). The balance of the future contract must include this deposit
+                // (v' in section 7 of yellowpaper).
+            })
+            .expect("There must be unused nonce");
+        info!("Account {id} minimal valid nonce is {nonce}");
+        (nonce.into(), weight)
+    }
+
+    /// Checks if such account exists in EVM.
+    fn is_known_to_evm(account_id: &H160) -> (bool, Weight) {
+        let has_code = AccountCodes::<R>::contains_key(account_id);
+        let weight = <R as frame_system::Config>::DbWeight::get().reads(1);
```

Comment:
Doesn't look it redundant to have `(bool, Weight)` for function signature when we always know that the function itself takes 1 read all time during execution? Can't we just properly calculate it in place where the function is called taking into account the function weight ?

Additionally, I would like to note that we can somewhere to do 1 READ for getting `account_codes` by `AccountCodes::<R>::get` and then use it without reading storage again

Reply:
Discussed it and decided that decomposition in this case would rather cause more confusion than good.